oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/facets/InsecureElasticSearchFacets.java

```diff
@@ -0,0 +1,103 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.jackrabbit.oak.plugins.index.elasticsearch.facets;
+
+import org.apache.jackrabbit.oak.plugins.index.elasticsearch.query.ElasticsearchSearcher;
+import org.apache.jackrabbit.oak.plugins.index.elasticsearch.query.ElasticsearchSearcherModel;
+import org.apache.jackrabbit.oak.plugins.index.elasticsearch.util.ElasticsearchAggregationBuilderUtil;
+import org.apache.jackrabbit.oak.plugins.index.search.spi.query.FulltextIndex;
+import org.apache.jackrabbit.oak.spi.query.QueryIndex;
+import org.elasticsearch.index.query.QueryBuilder;
+import org.elasticsearch.search.aggregations.Aggregation;
+import org.elasticsearch.search.aggregations.bucket.terms.Terms;
+import org.elasticsearch.search.aggregations.bucket.terms.TermsAggregationBuilder;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class InsecureElasticSearchFacets implements ElasticsearchFacets {
+    private static final Logger LOG = LoggerFactory.getLogger(InsecureElasticSearchFacets.class);
+
+    private ElasticsearchSearcher searcher;
+    private QueryBuilder query;
+    private QueryIndex.IndexPlan plan;
+    private ElasticsearchAggregationData elasticsearchAggregationData;
+
+    public InsecureElasticSearchFacets(ElasticsearchSearcher searcher, QueryBuilder query,
+                                       QueryIndex.IndexPlan plan, ElasticsearchAggregationData elasticsearchAggregationData) {
+        this.searcher = searcher;
+        this.query = query;
+        this.plan = plan;
+        this.elasticsearchAggregationData = elasticsearchAggregationData;
+    }
+
+    @Override
+    public Map<String, List<FulltextIndex.Facet>> getElasticSearchFacets(int numberOfFacets) throws IOException {
+        if (elasticsearchAggregationData != null && numberOfFacets == elasticsearchAggregationData.getNumberOfFacets()) {
+            return changeToFacetList(elasticsearchAggregationData.getAggregations().getAsMap());
+        }
+        LOG.warn("Facet data is being retrieved by again calling Elasticsearch");
+        List<TermsAggregationBuilder> aggregationBuilders = ElasticsearchAggregationBuilderUtil.getAggregators(plan, numberOfFacets);
+        ElasticsearchSearcherModel elasticsearchSearcherModel = new ElasticsearchSearcherModel.ElasticsearchSearcherModelBuilder()
+                .withQuery(query)
+                .withAggregation(aggregationBuilders)
+                .build();
+        Map<String, Aggregation> facetResult = searcher.search(elasticsearchSearcherModel).getAggregations().getAsMap();
+        return changeToFacetList(facetResult);
```

Comment:
I am wondering if this is needed at all. The query is exactly the same as the one in ElasticsearchResultRowIterator. Moreover, if the query returns 5 aggregated results but we ask for 10 we repeat the query,

Reply:
Changed equality condition with inequality, so that we won't call ES again if previously number of facets retrieved is greater that we are ask in this method.